### ISSUE TITLE
参数  n_ctx

### ISSUE THREAD

**Issue by haif-liu** [Morizeyao/GPT2-Chinese#56]
这个参数限定样本的长度。

但是，对于不同的篇章，长度是不一样的。而且多篇篇章通过[CLS][MASK]分隔后混在一起。

直接取长度为n_ctx的字符作为一个样本，感觉这样不太合理。

请问有什么考虑吗？

**Comment by Morizeyao** [comment]
GPT2是这么训练的。
我最早使用的方案是一篇文章一个样本，不用CLS MASK切割。文章如果长度没有超过n_ctx的话，就用PAD填充。
后来改用现在的方案，效果上差不多。

**Comment by haif-liu** [comment]
直接上觉得一个篇章或者一个段落或者一个句子为一样样本，再补全可能好一点吧。
如果补全的话，代码里是不是还要添加每个样本的实际长度数据？GPT2内部的模块都考虑了吗，还是说输入输出都需要自己再按样本的实际长度取值？

另外，如果要去词向量的话，是直接把里面的transformer的第一层取出来吗？

**Comment by haif-liu** [comment]
> GPT2是这么训练的。
> 我最早使用的方案是一篇文章一个样本，不用CLS MASK切割。文章如果长度没有超过n_ctx的话，就用PAD填充。
> 后来改用现在的方案，效果上差不多。

直接上觉得一个篇章或者一个段落或者一个句子为一样样本，再补全可能好一点吧。
如果补全的话，代码里是不是还要添加每个样本的实际长度数据？GPT2内部的模块都考虑了吗，还是说输入输出都需要自己再按样本的实际长度取值？

另外，如果要去词向量的话，是直接把里面的transformer的第一层取出来吗？

**Comment by haif-liu** [comment]
> GPT2是这么训练的。
> 我最早使用的方案是一篇文章一个样本，不用CLS MASK切割。文章如果长度没有超过n_ctx的话，就用PAD填充。
> 后来改用现在的方案，效果上差不多。

直接上觉得一个篇章或者一个段落或者一个句子为一样样本，再补全可能好一点吧。
如果补全的话，代码里是不是还要添加每个样本的实际长度数据？GPT2内部的模块都考虑了吗，还是说输入输出都需要自己再按样本的实际长度取值？

另外，如果要去词向量的话，是直接把里面的transformer的第一层取出来吗？

**Comment by haif-liu** [comment]
stride参数是尽量保证每个样本之间存在重合部分吗

这个参数取值有什么技巧吗？
取n_ctx一半吗

**Comment by Morizeyao** [comment]
嗯，直觉上是觉得一篇一个样本比较好的。
第二句我没太懂，补全PAD的时候肯定已经知道之前有几个token，以及最大长度nctx了，所以补几个PAD不需要添加额外数据吧。
“去词向量”是指取词向量？是的，可以取第一层之前的embedding的结果。

stride是尽量保证有重合，这个参数是我自己发明的...参数取值凭直觉。

**Comment by LeeCP8** [comment]
取半，前面n_ctx/2及后面n_ctx/2数据，只参与一次，而中间部分训练两次，建议取n_ctx-1,这样做，可以衔接数据，不然最后第n_ctx字与其下一个被切割。

**Comment by Morizeyao** [comment]
取nctx-1的话样本会特别多，当然可以是可以的

**Comment by LeeCP8** [comment]
是少很多的，移动距离更远了

**Comment by Morizeyao** [comment]
哦哦，我搞错了

**Comment by haif-liu** [comment]
> 嗯，直觉上是觉得一篇一个样本比较好的。
> 第二句我没太懂，补全PAD的时候肯定已经知道之前有几个token，以及最大长度nctx了，所以补几个PAD不需要添加额外数据吧。
> “去词向量”是指取词向量？是的，可以取第一层之前的embedding的结果。
> 
> stride是尽量保证有重合，这个参数是我自己发明的...参数取值凭直觉。

每个篇章作为一个样本的话，需要补全。这个是不是需要跟Bert一样，要做attention_mask？
GPT2里面这个参数是head_mask吗

**Comment by Morizeyao** [comment]
不是，pad部分没有做mask的机制

**Comment by haif-liu** [comment]
> 不是，pad部分没有做mask的机制

def forward(self, inputs, position_ids=None, token_type_ids=None, labels=None, past=None, head_mask=None):
        transformer_outputs = self.transformer(input_ids, position_ids=position_ids, token_type_ids=token_type_ids,
                                               past=past, head_mask=head_mask)
        hidden_states = transformer_outputs[0]

        lm_logits = self.lm_head(hidden_states)

        outputs = (lm_logits,) + transformer_outputs[1:]
        if labels is not None:
            # Shift so that tokens < n predict n
            shift_logits = lm_logits[..., :-1, :].contiguous()
            shift_labels = labels[..., 1:].contiguous()
            # Flatten the tokens
            loss_fct = CrossEntropyLoss(ignore_index=-1)
            loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)),
                            shift_labels.view(-1))
            outputs = (loss,) + outputs

        return outputs  # (loss), lm_logits, presents, (all hidden_states), (attentions)

GPT2LMHeadModel
forword里面
hidden_states = transformer_outputs[0]
为什么取得是第0个？这个表示的词向量那一层

**Comment by Morizeyao** [comment]
因为outputs是一个tuple，第一个元素是transformer部分的输出

**Comment by haif-liu** [comment]
> 因为outputs是一个tuple，第一个元素是transformer部分的输出

我是说0是词向量那一层，正常不应该是最后transformer最后一层吗

**Comment by Morizeyao** [comment]
这个0不是词向量那一层哦，是transformer的最后一层

**Comment by haif-liu** [comment]
> 这个0不是词向量那一层哦，是transformer的最后一层

还有一个问题。
语言模型的训练输入和输出是移位的，我看数据处理那块并没有做移位处理，输入和输出是一样的。

**Comment by Morizeyao** [comment]
模型里做了错位了，输入的时候不用再做一遍

**Comment by haif-liu** [comment]
> 模型里做了错位了，输入的时候不用再做一遍

好的，我再研究看看

**Comment by LeeCP8** [comment]
感谢两位朋友的问答，让我理清GPT内部处理的问题。
1 一般涉及NLP的任务时，我们总要处理长度问题，无论是基于句子，还是基于文章，长短都是不一致的。但是输入一般以三维向量为基准，为了提高处理效率，充分利用GPT或TPU的矩阵运算能力，一般都运行批量运算，使得输入要有一样的维度(max_seq_len)，这是需要解决长短不一致问题的根本原因。
2 当我们以句子为分割单位，进行批量输入时，我们要解决句子长度不一致问题，这时一般使用PAD来填充，然后使用一个相同维度的矩阵来声明PAD的位置，这样可以在进行Atten运算时，除去这些PAD，在误差计算时，除去这些PAD，不加入loss计算。而max_seq_len，这个是一个统一的长度，而我们的句子可长可短，长了我们依旧需要分割，短了需要补齐，所以无论是以句子或者文章为基本单位，最终还是得以max_seq_len为分割基准。对于文章同理。
3 所以我们根本没有必要纠结于，是应该以句子为单位还是文章为单位，最终还是以max_seq_len为基准。采用一个文章为样本，最终还是得分割，而且一个模型要具备通用性，就肯定要训练多篇文章，而且还要打乱才行，不打乱，以文章为单位，易出现过拟合。所以GPT采用的方式就是不用注意长度，直接加入标识符就行，以max_seq_len为基准分割，那么模型本身就会学习到这些特点。模型自己知道什么时候应该输出换行或者终止输出。
这时候，我们会有这样的问题，我训练模型，目的就是让模型产生文章或者句子，当然需要以文章或句子为单位才行啊，但是实际上模型是采用自回归的方式产生数据的，只要不限制长度，理论上可以无限产生数据，你想要句子，长度设置短一点就可以，如果想要文章，长度设置长一些就行。因为GPT模型始终能使用到的信息就是context-before(max_seq_len)。
4 那么这个时候又有一个问题，就是我不想自己设置长度，让模型自动生成句子或文章，那么训练时，添加标识符就行，产生时，判断标识符进行终止就可以了。下图是GPT接收外部数据的格式要求：
![snipaste_20190906_170350](https://user-images.githubusercontent.com/17487200/64425351-082d1480-d0de-11e9-8551-f7e512aada1f.png)
大家如果有什么见解，欢迎交流，谢谢！
